libraries/mpxe-store/src/store.c

```diff
@@ -0,0 +1,174 @@
+  struct pollfd pfd = { .fd = STDIN_FILENO, .events = POLLIN };
+  LOG_DEBUG("starting to poll\n");
+  while (true) {
+    int res = poll(&pfd, 1, -1);
+    if (res == -1) {
+      // interrupted
+      continue;
+    } else if (res == 0) {
+      continue;  // nothing to read
+    } else {
+      if (pfd.revents & POLLIN) {
+        static uint8_t buf[4096];  // store may be bigger than this?
+        int64_t len = read(STDIN_FILENO, buf, sizeof(buf));
+        if (len == -1) {
+          LOG_DEBUG("read error while polling\n");
+        }
+        prv_handle_store_update(buf, len);
+      } else {
+        LOG_DEBUG("pollhup\n");
+      }
+    }
+  }
+  return NULL;
+}
+
+void store_config(void) {
+  // do nothing after the first call
+  if (s_initialized) {
+    return;
+  }
+
+  // set up signal handler
+  signal(SIGUSR1, prv_sigusr);
+  pthread_mutex_init(&s_sig_lock, NULL);
+
+  // set up polling thread
+  pthread_t poll_thread;
+  pthread_create(&poll_thread, NULL, prv_poll_update, NULL);
+
+  // set up store pool
+  memset(&s_stores, 0, sizeof(s_stores));
+
+  // open fifo
+  char fifo_path[64];
+  snprintf(fifo_path, sizeof(fifo_path), "/tmp/%d_ctop", getpid());
+  mkfifo(fifo_path, 0666);
+  s_ctop_fifo = open(fifo_path, O_WRONLY);
+
+  s_initialized = true;
+}
+
+void store_register(MxStoreType type, StoreFuncs funcs, void *store, void *key) {
+  s_func_table[type] = funcs;
+  // malloc a proto as a store and return a pointer to it
+  Store *local_store = prv_get_first_empty();
+  if (store == NULL) {
+    return;
+  }
+  local_store->type = type;
+  local_store->store = store;
+  local_store->key = key;
+}
+
+void *store_get(MxStoreType type, void *key) {
+  // just linear search for it
+  // if key is NULL just get first one with right type
+  for (uint16_t i = 0; i < MAX_STORE_COUNT; i++) {
+    if (key == NULL && s_stores[i].type == type) {
+      return s_stores[i].store;
+    } else if (key != NULL && s_stores[i].type == type) {
+      if (s_stores[i].key == key) {
+        return s_stores[i].store;
+      }
+    }
+  }
+  return NULL;
+}
+
+void store_export(MxStoreType type, void *store, void *key) {
+  // Serialize store to proto
+  size_t packed_size = s_func_table[type].get_packed_size(store);
+  uint8_t *store_buf = malloc(packed_size);
+  s_func_table[type].pack(store, store_buf);
+
+  // Set up message sent to python
+  MxStoreInfo msg = MX_STORE_INFO__INIT;
+  msg.key = (uint64_t)key;
+  msg.type = type;
+  msg.msg.data = store_buf;
+  msg.msg.len = packed_size;
+
+  // Serialize export message
+  size_t export_size = mx_store_info__get_packed_size(&msg);
+  uint8_t *export_buf = malloc(export_size);
+  mx_store_info__pack(&msg, export_buf);
+
+  pthread_mutex_lock(&s_sig_lock);
+  // write proto to fifo
+  ssize_t written = write(s_ctop_fifo, export_buf, export_size);
+  // wait for signal that parent got message
+  pthread_mutex_lock(&s_sig_lock);
+  pthread_mutex_unlock(&s_sig_lock);
```

Comment:
using locks to wait to be signalled by the signal handler thread seems kinda hacky, can't think of a better way to do it though. Maybe a condition variable is a more natural way to do it? (https://web.stanford.edu/~ouster/cgi-bin/cs140-spring14/lecture.php?topic=locks.) Also, should there be a timeout mechanism here? Seems like it would block forever if the parent never signals for some reason, though maybe that's desired

Reply:
Condition variables require a mutex anyways, I think they're more useful for working with more threads (lol what do I know though) so I doubt we'd reduce much complexity. We could make it a timed wait but that also adds some more complexity in setting up the wait so this is probably fine for now.